internal/network/resource_network_forward.go

```diff
@@ -0,0 +1,466 @@
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+
+			"config": schema.MapAttribute{
+				Optional:    true,
+				Computed:    true,
+				ElementType: types.StringType,
+			},
+
+			"ports": schema.SetNestedAttribute{
+				Optional: true,
+				Computed: true,
+				Default:  setdefault.StaticValue(types.SetNull(portObjectType)),
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"description": schema.StringAttribute{
+							Optional:    true,
+							Description: "Port description",
+						},
+
+						"protocol": schema.StringAttribute{
+							Optional:    true,
+							Computed:    true,
+							Default:     stringdefault.StaticString("tcp"),
+							Description: "Port protocol",
+							Validators: []validator.String{
+								stringvalidator.OneOf("tcp", "udp"),
+							},
+						},
+
+						"listen_port": schema.StringAttribute{
+							Required:    true,
+							Description: "Listen port to forward",
+						},
+
+						"target_port": schema.StringAttribute{
+							Optional:    true,
+							Computed:    true,
+							Description: "Target port to forward listen port to. Defaults to the value of listen_port",
+						},
+
+						"target_address": schema.StringAttribute{
+							Required:    true,
+							Description: "Target address to forward listen port to",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func portObjectType() types.ObjectType {
+	return types.ObjectType{
+		AttrTypes: map[string]attr.Type{
+			"description":    types.StringType,
+			"protocol":       types.StringType,
+			"listen_port":    types.StringType,
+			"target_port":    types.StringType,
+			"target_address": types.StringType,
+		},
+	}
+}
+
+func (r *NetworkForwardResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	data := req.ProviderData
+	if data == nil {
+		return
+	}
+
+	provider, ok := data.(*provider_config.LxdProviderConfig)
+	if !ok {
+		resp.Diagnostics.Append(errors.NewProviderDataTypeError(req.ProviderData))
+		return
+	}
+
+	r.provider = provider
+}
+
+func (r *NetworkForwardResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	var config *NetworkForwardModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If any configured port has an empty value for target_port, use the same value as
+	// listen_port. This prevents "unknown value" in Terraform plan.
+	//
+	// Note: The modification is only needed because Terraform currently does not support
+	// dynamic default values.
+	if !req.Config.Raw.IsNull() {
+		ports, diags := ToNetworkForwardPortList(ctx, config.Ports)
+		if diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+
+		for i := range ports {
+			if ports[i].TargetPort == "" {
+				ports[i].TargetPort = ports[i].ListenPort
+			}
```

Comment:
Why not leave them unspecified? This way, LXD is the bit controlling what default gets applied.

Reply:
To show the user the resulting value in the Terraform plan and to ensure the value is always applied even if manually changed in LXD. We can also leave it unset.

Reply:
I'd prefer to leave it unset so that LXD's authoritative in the default behavior.